### PR TITLE
PR to fix Issue #1442 - don't parse pg interval values as JSON objects

### LINKED ISSUE #1442: BUG: INTERVAL columns receive wrong value when cloning row
**Describe the bug**
When `clone`ing a row where a column is of data type `INTERVAL(6)` and committing the new row without touching the field in the interval column, the value in the new row is changed to the shortes (?) possible interval given the original interval (truncating all zeros on the right maybe?).

**To Reproduce**
- Connect to a postgres database (14 in my case, but I think I remember this from using an older postgres version as well)
- create a table where one column is of type `interval`, e.g.
```
CREATE TABLE
  automation.ingestion_config (
    sensor_id uuid NOT NULL,
    mqtt_topic text NOT NULL,
    persist boolean NOT NULL DEFAULT true,
    target_table text NULL,
    expected_frequency interval NULL
  );

ALTER TABLE
  automation.ingestion_config
ADD
  CONSTRAINT ingestion_config_pkey PRIMARY KEY (mqtt_topic)
```
- insert a row in the table
- open the table, right-click on the row and select `clone row`
- (change/set new value for primary key column)
- click `Apply`
- check the new value in the `interval`-type column

**Expected behavior**
the cloned row should have the same value as the source row. 
It seems to me all zeros on the right are removed, e.g. a source value of `00:10:00` (10 minutes) is turned into `00:00:01` (1 second), while a source value of `00:15:00` (15 minutes) is turned into `00:00:15` (15 seconds)...

**Version Information (please complete the following information):**
 - OS: Ubuntu
 - App Version 3.7.2
 - PostgreSQL 14.5 (Ubuntu 14.5-0ubuntu0.22.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0, 64-bit

### PR DESCRIPTION
**Pull request by davidkaufman**
Treat pg's interval columns similar to the way we handle date, datetime and timestamp columns
so that they can be used as literal values for updates & inserts without re/un-conversion

Fix #1442

### CODE REVIEW

**Review by rathboma — CHANGES_REQUESTED** [review]
Nice. Crazy detailed write-up!

Can you add a unit test or two to validate as working?

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 1]
```diff
@@ -47,11 +47,17 @@ function tableName(table: string, schema?: string): string{
...
+ * See also: https://github.com/brianc/node-postgres/issues/285
+ * (and note that the code refrenced in /lib/textParsers.js has been broken out into it own module
+ * so it now lives in https://github.com/brianc/node-pg-types/blob/master/lib/textParsers.js#L175)
+ *
  * TODO: do not convert as well these same types with array (types 1115, 1182, 1185)
  */
 pg.types.setTypeParser(1082, 'text', (val) => val); // date
 pg.types.setTypeParser(1114, 'text', (val) => val); // timestamp without timezone
 pg.types.setTypeParser(1184, 'text', (val) => val); // timestamp
+pg.types.setTypeParser(1186, 'text', (val) => val); // interval (Issue #1442 "BUG: INTERVAL columns receive wrong value when cloning row)
```
Comment (MasterOdin) [inline apps/studio/src/lib/db/clients/postgresql.ts:60]: The `pg.types` object provides a `builtins` object that has constants for these numbers which might be easier to use:

```
pg.types.setTypeParser(pg.types.builtins.DATE, (val) => val);
pg.types.setTypeParser(pg.types.builtins.TIMESTAMP, (val) => val);
pg.types.setTypeParser(pg.types.builtins.TIMESTAMPTZ, (val) => val);
pg.types.setTypeParser(pg.types.builtins.INTERVAL, (val) => val);
```

Comment (davidkaufman) [reply apps/studio/src/lib/db/clients/postgresql.ts:60]: Thanks, @MasterOdin! Definitely will do that

**Comment by davidkaufman** [comment]
@rathboma wrote
> Nice. Crazy detailed write-up!

yeah I may be a closet/frustrated technical writer -- if so, its both a superpower and an achilles heel 😃

> Can you add a unit test or two to validate as working?

Yes! I thought.  Of course I can.  This being the first time running the pg integration tests tho... turned out easier thought than done.  Seems all 24 (of the 96) pg.spec tests fail the same way:

~~~bash
  ● Postgres [latest - socket? true] › Should allow me to update rows with an empty array

    There is no connection available.

      482 | function checkIsConnected(_server: IDbConnectionServer, database: IDbConnectionDatabase) {
      483 |   if (database.connecting || !database.connection) {
    > 484 |     throw new Error('There is no connection available.');
          |           ^
      485 |   }
      486 | }
      487 |

      at checkIsConnected (src/lib/db/client.ts:484:11)
      at applyChanges (src/lib/db/client.ts:359:3)
      at Object.<anonymous> (tests/integration/lib/db/clients/postgres.spec.ts:97:44)
          at runMicrotasks (<anonymous>)

    connect ENOENT /var/folders/8x/cg3k1p_j09b4qb38kyj33my80000gq/T/psql-i2fNUW/postgresql/.s.PGSQL.5432
~~~

I can see (in Docker Desktop) the earlier tests creating pg containers and their tests working, so I'm digging into this to see why.  Do these run successfully for you on Ubuntu @rathboma?

**Comment by davidkaufman** [comment]
And it loks like all the failing tests are the `socket? true` ones:
~~~python
  Postgres [9.3 - socket? false]
    ✓ Should allow me to update rows with an empty array (48ms)
    ✓ Should allow me to insert a row with an array (40ms)
    ✓ Should allow me to update rows with array types (43ms)
    Common Tests
      ✓ get database version should work (1ms)
      ✓ list tables should work (21ms)
      ✓ column tests (180ms)
      ✓ table view tests (68ms)
      ✓ stream tests (58ms)
      ✓ query tests (23ms)
      ✓ get insert query tests (13ms)
      ✓ table filter tests (17ms)
      ✓ table triggers (13ms)
      ✓ primary key tests (30ms)
      Alter Table Tests
        ✓ should past alter table tests (139ms)
        ✓ should alter indexes (93ms)
      Table Structure
        ✓ should fetch table properties (94ms)
      Data modification
        ✓ should insert good data (78ms)
        ✓ should not insert bad data (136ms)
        ✓ should apply all types of changes (78ms)
        ✓ should not commit on change error (107ms)
      Data modification (with composite primary key)
        ✓ should insert good data (68ms)
        ✓ should not insert bad data (67ms)
        ✓ should apply all types of changes (87ms)
        ✓ should not commit on change error (96ms)
  Postgres [9.4 - socket? false]
    ✓ Should allow me to update rows with an empty array (20ms)
    ✓ Should allow me to insert a row with an array (33ms)
    ✓ Should allow me to update rows with array types (23ms)
    Common Tests
      ✓ get database version should work (1ms)
      ✓ list tables should work (15ms)
      ✓ column tests (130ms)
      ✓ table view tests (64ms)
      ✓ stream tests (41ms)
      ✓ query tests (13ms)
      ✓ get insert query tests (10ms)
      ✓ table filter tests (13ms)
      ✓ table triggers (9ms)
      ✓ primary key tests (22ms)
      Alter Table Tests
        ✓ should past alter table tests (105ms)
        ✓ should alter indexes (70ms)
      Table Structure
        ✓ should fetch table properties (66ms)
      Data modification
        ✓ should insert good data (72ms)
        ✓ should not insert bad data (56ms)
        ✓ should apply all types of changes (63ms)
        ✓ should not commit on change error (82ms)
      Data modification (with composite primary key)
        ✓ should insert good data (58ms)
        ✓ should not insert bad data (55ms)
        ✓ should apply all types of changes (82ms)
        ✓ should not commit on change error (86ms)
  Postgres [latest - socket? false]
    ✓ Should allow me to update rows with an empty array (30ms)
    ✓ Should allow me to insert a row with an array (45ms)
    ✓ Should allow me to update rows with array types (32ms)
    Common Tests
      ✓ get database version should work (1ms)
      ✓ list tables should work (25ms)
      ✓ column tests (183ms)
      ✓ table view tests (62ms)
      ✓ stream tests (48ms)
      ✓ query tests (16ms)
      ✓ get insert query tests (13ms)
      ✓ table filter tests (15ms)
      ✓ table triggers (9ms)
      ✓ primary key tests (116ms)
      Alter Table Tests
        ✓ should past alter table tests (161ms)
        ✓ should alter indexes (86ms)
      Table Structure
        ✓ should fetch table properties (374ms)
      Data modification
        ✓ should insert good data (82ms)
        ✓ should not insert bad data (69ms)
        ✓ should apply all types of changes (78ms)
        ✓ should not commit on change error (112ms)
      Data modification (with composite primary key)
        ✓ should insert good data (70ms)
        ✓ should not insert bad data (70ms)
        ✓ should apply all types of changes (103ms)
        ✓ should not commit on change error (97ms)
  Postgres [latest - socket? true]
    ✕ Should allow me to update rows with an empty array (52ms)
    ✕ Should allow me to insert a row with an array (1ms)
    ✕ Should allow me to update rows with array types
    Common Tests
      ✕ get database version should work (20ms)
      ✕ list tables should work (1ms)
      ✕ column tests
      ✕ table view tests (1ms)
      ✕ stream tests (201ms)
      ✕ query tests (1ms)
      ✕ get insert query tests (1ms)
      ✕ table filter tests (1ms)
      ✕ table triggers (1ms)
      ✕ primary key tests (1ms)
      Alter Table Tests
        ✕ should past alter table tests (129ms)
        ✕ should alter indexes (175ms)
      Table Structure
        ✕ should fetch table properties (1ms)
      Data modification
        ✕ should insert good data (111ms)
        ✕ should not insert bad data (115ms)
        ✕ should apply all types of changes (87ms)
        ✕ should not commit on change error (111ms)
      Data modification (with composite primary key)
        ✕ should insert good data (59ms)
        ✕ should not insert bad data (89ms)
        ✕ should apply all types of changes (88ms)
        ✕ should not commit on change error (113ms)

~~~

**Comment by davidkaufman** [comment]
(feel free to telly to just tell me to shut up and code, if i'm going crazy detailed again :-)

**Comment by davidkaufman** [comment]
@rathboma jus to clarify, you asked for a "unit test" or two, but to prove this bug, I need to add test to the integration test suite because the unit tes suite doesn't have access to an actual posgres db test connection right?

**Comment by davidkaufman** [comment]
I having a hard time recreating the behavior reported (inserting a cloned interval of minutes, and having it come back out of the db as seconds) using db client libs in test code; The bug is easy to reproduce clicking on the ui, but infuriatingly not broken using `knex` or `pg` calls; The behavior appears to be somewhere in the ui code (Tabulator, maybe?) and not sure how to write a test for that (without going full selenium webdriver on it...)

**Comment by davidkaufman** [comment]
things that don't work (or rather, __do__ work, so can't be used to prove the bug):

~~~ks
await util.knex("test_intervals").insert( clonedRow )
~~~

~~~js
await util.connection.applyChanges({ inserts: [ { ...clonedRow } ] })
~~~
# 🧐

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 2]
```diff
@@ -47,11 +48,17 @@ function tableName(table: string, schema?: string): string{
...
+ * See also: https://github.com/brianc/node-postgres/issues/285
+ * (and note that the code refrenced in /lib/textParsers.js has been broken out into it own module
+ * so it now lives in https://github.com/brianc/node-pg-types/blob/master/lib/textParsers.js#L175)
+ *
  * TODO: do not convert as well these same types with array (types 1115, 1182, 1185)
  */
 pg.types.setTypeParser(1082, 'text', (val) => val); // date
 pg.types.setTypeParser(1114, 'text', (val) => val); // timestamp without timezone
 pg.types.setTypeParser(1184, 'text', (val) => val); // timestamp
+// pg.types.setTypeParser(1186, 'text', (val) => val); // interval (Issue #1442 "BUG: INTERVAL columns receive wrong value when cloning row)
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/postgresql.ts:61]: @davidkaufman looks like some testing changes here need to be reverted before I merge it?

Comment (davidkaufman) [reply apps/studio/src/lib/db/clients/postgresql.ts]: Oh, duh yes! fixed that :-)

And -- yay! 😁  I simplified the test and finally got it to detect the problem!

~~~patch
dkaufman@bender studio % yarn test:integration --runInBand --verbose postgres -t issue-1442                                                          yarn run v1.22.19
$ TEST_MODE=1 ELECTRON_RUN_AS_NODE=1 yarn electron ../../node_modules/jest/bin/jest.js --config ./jest.integration.config.js --runInBand --verbose postgres -t issue-1442
$ /Users/dkaufman/dev/bks/beekeeper-studio-fork/node_modules/.bin/electron ../../node_modules/jest/bin/jest.js --config ./jest.integration.config.js --runInBand --verbose postgres -t issue-1442

  console.info tests/integration/lib/db/clients/postgres.spec.ts:161
+++ inserted data:  { id: 1, amount_of_time: '00:15:00' }

  console.log tests/integration/lib/db/clients/postgres.spec.ts:168
+++ retrieved data:  { id: 1, amount_of_time: PostgresInterval { minutes: 15 } }

 FAIL  tests/integration/lib/db/clients/postgres.spec.ts (24.422s)
  Postgres [9.4 - socket? false]
    ✕ Should clone interval values in pg-intervalStyle format not json (issue-1442) (32ms)

  ● Postgres [9.4 - socket? false] › Should clone interval values in pg-intervalStyle format not json (issue-1442)

    expect(received).toStrictEqual(expected) // deep equality

- Expected
+ Received

  Object {
-   "amount_of_time": "00:15:00",
+   "amount_of_time": PostgresInterval {
+     "minutes": 15,
+   },
    "id": 1,
  }

      169 |
      170 |       // retrieved interval value should be the same, unchanged interval (string) "00:15:00"
    > 171 |       expect ( retrievedData ).toStrictEqual({
          |                                ^
      172 |         id: 1,
      173 |         amount_of_time: insertedValue // should still be a string not an object
      174 |       })

      at Object.<anonymous> (tests/integration/lib/db/clients/postgres.spec.ts:171:32)
          at runMicrotasks (<anonymous>)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 24 skipped, 25 total
Snapshots:   0 total
Time:        25.943s, estimated 52s
Ran all test suites matching /postgres/i with tests matching "issue-1442".
error Command failed with exit code 1.
~~~

Not sure what i was doing wrong before ...probably an insufficient coffee dosage

Also implemented @MasterOdin's suggestion to replace all those obscure datatype id's with the `pg.types.builtins` constants

**Comment by davidkaufman** [comment]
Oops sorry, it looks like i merged when i should have rebased this branch https://github.com/beekeeper-studio/beekeeper-studio/pull/1465/commits/e9130bb34d2d932708628622f622752584ae9421

Let me know if that makes it harder (or messier) for you to merge into master, and I'll recreate a new clean branch and submit another PR!

**Review by rathboma — APPROVED** [review]

**Pull request merged**